nomad/structs/structs.go

```diff
@@ -6432,12 +6447,67 @@ func (a *AllocMetric) ExhaustQuota(dimensions []string) {
 	a.QuotaExhausted = append(a.QuotaExhausted, dimensions...)
 }
 
+// ScoreNode is used to gather top K scoring nodes in a heap
 func (a *AllocMetric) ScoreNode(node *Node, name string, score float64) {
-	if a.Scores == nil {
-		a.Scores = make(map[string]float64)
+	if a.topScores == nil {
+		a.topScores = make(map[string]*lib.ScoreHeap)
+	}
+	scorerHeap, ok := a.topScores[name]
+	if !ok {
+		scorerHeap = lib.NewScoreHeap(MaxRetainedNodeScores)
+		a.topScores[name] = scorerHeap
+	}
+	heap.Push(scorerHeap, &lib.HeapItem{
+		Value: node.ID,
+		Score: score,
+	})
+}
+
+// PopulateScoreMetaData populates a map of scorer to scoring metadata
+// The map is populated by popping elements from a heap of top K scores
+// maintained per scorer
+func (a *AllocMetric) PopulateScoreMetaData() {
+	if a.ScoreMetaData == nil {
+		a.ScoreMetaData = make(map[string][]*NodeScoreMeta)
+	}
+	if a.topScores != nil && len(a.ScoreMetaData) == 0 {
+		for scorer, scoreHeap := range a.topScores {
+			scoreMeta := make([]*NodeScoreMeta, scoreHeap.Len())
+			i := scoreHeap.Len() - 1
+			// Pop from the heap and store in reverse to get top K
+			// scoring nodes in descending order
+			for scoreHeap.Len() > 0 {
+				item := heap.Pop(scoreHeap).(*lib.HeapItem)
+				scoreMeta[i] = &NodeScoreMeta{
```

Comment:
Is it worth just adding a GetItems() and GetItemsReverse() method to the library?